File: test/four/app/src/main/java/com/z3d/display/line/LineDisplaySprite.java
package com.z3d.display.line;

import android.opengl.GLES20;
import android.util.Log;

import com.z3d.base.ObjData;
import com.z3d.core.Context3D;
import com.z3d.display.basedis.DisplayBaseSprite;
import com.z3d.scene.Scene3D;
import com.z3d.vo.Vector3D;

import java.util.ArrayList;

public class LineDisplaySprite extends DisplayBaseSprite {

    public Vector3D baseColor;
    LineDisplaySprite(Scene3D val){
        super(val);
        scene3D.progrmaManager.registe(LineDisplayShader.shaderNameStr,new LineDisplayShader(scene3D));
        this.shader3D=this.scene3D.progrmaManager.getProgram(LineDisplayShader.shaderNameStr);
    }


    protected void  initData()
    {
        baseColor=new Vector3D(1.0f,0.0f,1.0f);
        this.objData =new ObjData(scene3D);
        this.clearLine();


        this.makeLineMode(new Vector3D(0,0 ,0) ,new Vector3D(500,0,0));
        this.makeLineMode(new Vector3D(0,0 ,0) ,new Vector3D(0,100,0));

        this.upLineDataToGpu();
    }

    public void  upLineDataToGpu()
    {
        this.objData.upToGup();
    }
    public  void clearLine()
    {

        this.objData.verticeslist=new ArrayList<Float>();//结果顶点坐标列表
        this.objData.normals=new ArrayList<Float>();//结果顶点坐标列表
        this.objData.indexs=new ArrayList<Short>();
    }
    public  void  makeLineMode(Vector3D a, Vector3D b )
    {
        this.makeLineMode(a, b,this.baseColor);
    }
    public  void  makeLineMode(Vector3D a, Vector3D b, Vector3D c)
    {
        this.objData.verticeslist.add(a.x );
        this.objData.verticeslist.add(a.y );
        this.objData.verticeslist.add(a.z );

        this.objData.verticeslist.add(b.x );
        this.objData.verticeslist.add(b.y );
        this.objData.verticeslist.add(b.z );


        this.objData.normals.add(c.x );
        this.objData.normals.add(c.y );
        this.objData.normals.add(c.z );

        this.objData.normals.add(c.x );
        this.objData.normals.add(c.y );
        this.objData.normals.add(c.z );


        this.objData.indexs.add((short) this.objData.indexs.size());
        this.objData.indexs.add((short) this.objData.indexs.size());
        this.objData.isCompile=false;
    }


    public void upData(){


        if(this.shader3D!=null&&this.objData.treNum>0){
            Context3D ctx=this.scene3D.context3D;
            this.modeMatrix.appendRotation(1, Vector3D.Z_AXIS);
            ctx.setProgame(this.shader3D.program);



            ctx.setVcMatrix4fv(this.shader3D,"vpMatrix3D",this.scene3D.camera3D.modelMatrix.m);
            ctx.setVcMatrix4fv(this.shader3D,"posMatrix",this.modeMatrix.m);

            ctx.setVa(this.shader3D,"vPosition",3,this.objData.vertexBuffer);
            ctx.setVa(this.shader3D,"vColorv3d",3,this.objData.normalsBuffer);

            GLES20.glDrawElements(GLES20.GL_LINES,this.objData.treNum, GLES20.GL_UNSIGNED_SHORT,this.objData.indexBuffer);

        }


    }
}

File: new/src/Pan3d/display/line/LineDisplaySprite.ts
module Pan3d {
   
    export class LineDisplaySprite  extends Display3D  {
        private shader3D:Shader3D;
        private objData: ObjData;
        private lineVecPos: Array<number>;
        private lineColor: Array<number>
        private lineIndex: Array<number>
        protected baseColor: Vector3D = new Vector3D(1, 0, 0)
        constructor(value: Scene3D) {
          super(value);
          this.scene3D.progrmaManager.registe(LineDisplayShader.LineDisplayShader,new LineDisplayShader(this.scene3D));
          this.shader3D=  this.scene3D.progrmaManager.getProgram(LineDisplayShader.LineDisplayShader);
          this.initData();
    
        }
        protected initData():void
        {

            this.makeLineMode(new Vector3D(0,0,0),new Vector3D(100,0,0),new Vector3D(1,0,0));
            this.upToGpu();
        }
        protected upToGpu():void
        {
            this.objData.upToGpu();
        }
        protected makeLineMode(a: Vector3D, b: Vector3D, $color: Vector3D = null): void {
            if (!this.lineVecPos || !this.lineIndex) {
                this.clear()
            }
            if ($color) {
                this.baseColor = $color
            }
            this.lineVecPos.push(a.x, a.y, a.z);
            this.lineVecPos.push(b.x, b.y, b.z);
            this.lineColor.push(this.baseColor.x, this.baseColor.y, this.baseColor.z)
            this.lineColor.push(this.baseColor.x, this.baseColor.y, this.baseColor.z)
            this.lineIndex.push(this.lineIndex.length + 0, this.lineIndex.length + 1)
            if(this.objData==null){
                this.objData=new ObjData(this.scene3D);
            }
            this.objData.treNum = this.lineIndex.length
            this.objData.vertices =  this.lineVecPos;
            this.objData.normals = this.lineColor;
            this.objData.indexs =  this.lineIndex;

        }
     
        public clear(): void {
            this.lineVecPos = new Array;
            this.lineIndex = new Array
            this.lineColor = new Array
 
        }

        public upFrame(): void {
            if (this.objData && this.objData.indexBuffer) {
                var context3D:Context3D=this.scene3D.context3D;
                context3D.setProgram(this.shader3D.program);
                context3D.setVa(0, 3, this.objData.vertexBuffer);
                context3D.setVa(1, 3, this.objData.normalsBuffer);
                context3D.setVcMatrix4fv(this.shader3D, "vpMatrix3D", this.scene3D.camera3D.modelMatrix.m);
                context3D.setVcMatrix4fv(this.shader3D, "posMatrix", this.posMatrix.m);
                context3D.drawLine(this.objData.indexBuffer, this.objData.treNum);
            
            }
        }
        



    }
}